ImputationPipeline/Imputation.wdl

```diff
@@ -0,0 +1,453 @@
+      input_vcfs = array_vcfs,
+      input_vcf_indices = array_vcf_indices,
+      output_vcf_basename = "merged_aou_arrays"
+  }
+
+  if (perform_qc_steps) {
+        call QConArray {
+          input:
+            input_vcf = GenerateDataset.output_vcf,
+            input_vcf_index = GenerateDataset.output_vcf_index,
+            output_vcf_basename = "merged_and_QCd",
+        }
+  }
+
+  File to_be_imputed_vcf = select_first([QConArray.output_vcf, GenerateDataset.output_vcf]) 
+  File to_be_imputed_vcf_index = select_first([QConArray.output_vcf_index, GenerateDataset.output_vcf_index])
+
+  scatter (chrom in ["1","2","3","4","5","6","7","8","9","10","11","12","13","14","15","16","17","18","19","20","21","22"]) { 
+    call CalculateChromsomeLength {
+      input:
+        ref_dict = ref_dict,
+        chrom = chrom
+    }
+    
+    Float chunkLengthFloat = chunkLength
+    Int num_chunks = ceil(CalculateChromsomeLength.chrom_length / chunkLengthFloat)
+
+    scatter (i in range(num_chunks)) {
+      call ChunkVCF {
+        input:
+          start = (i * chunkLength) + 1,
+          end = if (CalculateChromsomeLength.chrom_length < ((i + 1) * chunkLength)) then CalculateChromsomeLength.chrom_length else ((i + 1) * chunkLength), 
+          vcf = to_be_imputed_vcf,
+          vcf_index = to_be_imputed_vcf_index,
+          chrom = chrom,
+          basename = "chrom_" + chrom + "_chunk_" + i
+       }
+
+      call QConChunk {
+        input: 
+          vcf = ChunkVCF.output_vcf,
+          vcf_index = ChunkVCF.output_vcf_index,
+          panel_vcf = path_to_reference_panel + "ALL.chr" + chrom + ".phase3_integrated.20130502.genotypes.vcf.gz",
+          panel_vcf_index = path_to_reference_panel + "ALL.chr" + chrom + ".phase3_integrated.20130502.genotypes.vcf.gz.tbi"
+      }
+
+      if (QConChunk.valid) {
+
+      call PrePhaseVariantsEagle {
+        input:
+          dataset_bcf = QConChunk.valid_chunk_bcf,
+          dataset_bcf_index = QConChunk.valid_chunk_bcf_index,
+          reference_panel_bcf = path_to_reference_panel + "ALL.chr" + chrom + ".phase3_integrated.20130502.genotypes.bcf",
+          reference_panel_bcf_index = path_to_reference_panel + "ALL.chr" + chrom + ".phase3_integrated.20130502.genotypes.bcf.csi",
+          chrom = chrom,
+          genetic_map_file = genetic_maps_eagle,
+          start = (i * chunkLength) + 1,
+          end = (i + 1) * chunkLength + 5000000 # they do an overlap of 5,000,000 bases in michigan pipeline
+      }
+
+        call minimac4 {
+          input:
+            ref_panel = path_to_m3vcf + chrom + ".1000g.Phase3.v5.With.Parameter.Estimates.m3vcf.gz",
+            phased_vcf = PrePhaseVariantsEagle.dataset_prephased_vcf,
+            prefix = "chrom" + "_chunk_" + i +"_imputed",
+            chrom = chrom,
+            start = (i * chunkLength) + 1,
+            end = (i + 1) * chunkLength
+        }
+      }
+     }
+  }
+
+  Array[File] phased_vcfs = flatten(select_all(minimac4.vcf))
```

Comment:
@kachulis : how would I get this to work? I'm getting this error now:

`Failed to process workflow definition 'TestingNewImputation' (reason 1 of 1): Failed to process declaration 'Array[File] phased_vcfs = flatten(select_all(minimac4.vcf))' (reason 1 of 1): Failed to process expression 'flatten(select_all(minimac4.vcf))' (reason 1 of 1): Invalid parameter 'IdentifierMemberAccess(minimac4,vcf,Vector())'. Expected an array of optional values (eg 'Array[X?]') but got 'Array[Array[File?]]'`

Reply:
oh actually I got it to work now ( I think )